### DISCUSSION TITLE
Movement jitter

### DISCUSSION THREAD

**Discussion by jeremyhcho** [foxssake/netfox#212] [Q&A]
Apologies for bombarding the Q&A today but I've seemingly got things to work except interpolation and I'm not sure what I'm missing.

Here's my `RollbackSynchronizer`
![image](https://github.com/foxssake/netfox/assets/13476975/46eb4e02-d261-444c-a0a9-d5b641a9e99a)

My `TickInterpolator`
![image](https://github.com/foxssake/netfox/assets/13476975/25e8a390-4515-4b55-9c05-989e55315e8b)

And my `_rollback_tick` code
```
func _rollback_tick(delta: float, _tick: int, _is_fresh: bool) -> void:
    # Add the gravity.
    _force_update_is_on_floor()
    if not is_on_floor():
        velocity.y -= gravity * delta

    var direction = Vector3(input.input_dir.x, 0, input.input_dir.y).normalized()
    if direction:
        velocity.x = direction.x * speed
        velocity.z = direction.z * speed
    else:
        velocity.x = move_toward(velocity.x, 0, speed)
        velocity.z = move_toward(velocity.z, 0, speed)

    # Apply movement
    velocity *= NetworkTime.physics_factor
    move_and_slide()
    velocity /= NetworkTime.physics_factor
 ```

With this setup I can move my character on the appropriate clients as well as see the correct characters move on opposite clients but the model jitters quite heavily during movement. According to the docs (unless I missed something right in front of me), this setup seems reasonable and would indicate that I should see smooth movement. Eyeball test indicates that the `TickInterpolator` does _something_ because when I turn off the `Enabled` flag it feels different but it's somewhat hard to tell.

I would appreciate a nudge in the right direction! Thanks in advance!

**Comment by elementbound** [comment]
Hi @jeremyhcho!

This kind of jitter usually happens when there's a difference in what the client predicts and what the server responds with. The jitter itself is the client adjusting its state, which is usually not much of an issue, unless a lot of adjustment is needed. 

Would you be able to check if input is properly received on the server? For example that the Input node's values approximately correspond to what you're pressing on the keyboard. 

Also, would you be able to share a minimum reproducing project? I'd be happy to look into it. Otherwise your setup looks good. 

> Eyeball test indicates that the TickInterpolator does something

Protip: you can test and see the TickInterpolator in action by setting the tickrate to something really low, like 8 or 4.

**Reply by jeremyhcho** [reply]
Thanks for the response!

I believe I confirmed that the input the server is reading is aligned with what I'm expecting. In terms of the client adjusting its state - that should only occur when the client's prediction is incorrect right?

I did briefly take a look at the debugging profiler and there didn't seem to be an alarming number of rollbacks either. If the client was adjusting its state that would trigger a rollback right? Or am I misunderstanding

**Reply by jeremyhcho** [reply]
Ah interesting so I'm seeing that the velocity on the server is quite different than the velocity on the client

**Reply by jeremyhcho** [reply]
I'm not sure how relevant that is but yeah it seems like `velocity` right before I multiply by `NetworkTime.physics_factor` is the same but right after the server evaluates it as roughly `-29` and the client evaluates it as `-24`

**Reply by jeremyhcho** [reply]
One thing is that I run my server through the command line `godot --headless --server`

**Reply by jeremyhcho** [reply]
Hmm I probably don't have a great understanding of this but the process delta is different on my server vs. client and that's what's causing the `velocity` to be a different number on the respective instances. Maybe that's ok?

**Reply by jeremyhcho** [reply]
On second look as well it almost looks like turning off the `TickInterpolator` makes it less jittery somehow. I also tried with a lower tick rate with `TickInterpolator` enabled and it makes the jitter a lot clearer.


https://github.com/foxssake/netfox/assets/13476975/69222852-2d5f-4ee8-8469-3cbec96a37ef

This video doesn't _exactly_ capture what I'm seeing on my end but essentially I'm seeing the other character flicker on every frame/tick.

**Reply by jeremyhcho** [reply]
Oh it looks like it fixed itself once I checked the "Sync to physics" box under settings. Is this not a recommended setting?

**Reply by elementbound** [reply]
> that should only occur when the client's prediction is incorrect right?

Correct!

> If the client was adjusting its state that would trigger a rollback right?

Not currently - rollback always happens from the latest state received from the server. So, by extension, the number of rollback ticks depends on the latency, not on the amount of adjustment needed to be done.

> I'm not sure how relevant that is but yeah it seems like velocity right before I multiply by NetworkTime.physics_factor is the same but right after the server evaluates it as roughly -29 and the client evaluates it as -24

This is a good clue! This means that the `physics_factor` is different on the two instances of the game. Physics factor is calculated based on Godot's physics tickrate setting and netfox's network tickrate setting. So one of those must be different.

> Hmm I probably don't have a great understanding of this but the process delta is different on my server vs. client and that's what's causing the velocity to be a different number on the respective instances. Maybe that's ok?

By process delta, do you mean the delta you receive in the _process callback, or in one of netfox's callbacks? If the _process delta differs that's fine, if netfox's delta differs that's an issue.

> it fixed itself once I checked the "Sync to physics" box under settings. Is this not a recommended setting?

It's completely fine to use! That settings ties the network tick loop to Godot's physics updates. I've added it as an option in case you need to tie the networked game logic closer to the physics system, so this way you can run both in the same callback.

---

So based on the above ( differing physics factor, sync to physics fixing the issue ), I have a strong hunch that somehow the tickrate was different on the client and the server. Would you be able to confirm that, e.g. by printing the peer id, `NetworkTime.tickrate`, and the delta in either `NetworkTime.on_tick` and/or `_rollback_tick`?

I'm wondering if there's some kind of bug where netfox doesn't pick up the right config when running a headless build.

**Reply by jeremyhcho** [reply]
Thanks for the response @elementbound!

I added print statements for `NetworkTime.on_tick`, `_rollback_tick` and printed `NetworkTime.tickrate` and they all seem to be equivalent. Here's what I'm seeing on both server and client:
```
NetworkTime.tickrate: 24
Delta in _rollback_tick: 0.04166666666667
Delta in on_tick: 0.04166666666667
```

**Reply by albertok** [reply]
Something that can make this happen is if you have movement logic of any kind in _process or _physics_process.

The fact that enabling  "Sync to physics" fixed things makes me strongly suspect it.

@jeremyhcho Can you double check absolutely nothing at all is manipulating positions or velocity in those?

**Reply by jeremyhcho** [reply]
Thanks for the suggestion @albertok

I went back and looked at all of my `_process` and `_process_physics` definitions and none of them are performing any sort of movement on the character.

- My `_process` function in `Player` is updating animation blend variables and `_physics_process` is undefined
- My `PlayerCamera` has no `_process` nor `_physics_process` defined
- My `PlayerInput` node has somewhat relevant functions that look like:
```
func _process(_delta: float) -> void:
	if Input.is_action_just_pressed("stab"):
		is_stabbing = true

	if Input.is_action_just_pressed("jump"):
		is_jumping = true

func _gather():
	input_dir = Input.get_vector(
		"move_left",
		"move_right",
		"move_forward",
		"move_backward"
	)
```

But neither of those functions look like offenders. There's nothing else in my code that looks at movement so I'm not sure what else to check :(

**Reply by jeremyhcho** [reply]
Following up here. Any other suspicions as to what might be causing this? cc @elementbound

I've been experimenting with the `Sync to physics` option checked but this sets the `NetworkTime.tickrate` to 60 which is higher than I'd need. On top of that, when I have this checked, there's something that seems to be a race condition with my player input component that looks like
```
[WRN][netfox::RollbackSynchronizer] Input:is_jumping
[WRN][netfox::RollbackSynchronizer] Received input for node owned by 2135926936 from 1, sender has no authority!
```

I'd like to try and fix the jitter without checking `Sync to physics`. Here's the current state of things:

`Player` scene tree
![image](https://github.com/foxssake/netfox/assets/13476975/952da760-3bac-4940-95a9-bf048dfedb91)

`RollbackSynchronizer`
![image](https://github.com/foxssake/netfox/assets/13476975/b872c259-1c6c-41df-87b4-2dd588aaaf78)

`TickInterpolator`
![image](https://github.com/foxssake/netfox/assets/13476975/a195e883-3ff8-4db4-b0e9-aa7de13e3a18)


`Player.ready()`
```
func _ready() -> void:
	# Prevent the spring from detecting the character model as an obstacle
	spring.add_excluded_object(self)

	global_position = GameStart.DEFAULT_SPAWN
	yaw.global_position = Vector3(
		global_position.x,
		yaw.global_position.y,
		global_position.z
	)

	# Process a frame before processing settings
	await get_tree().process_frame

	# Set authorities
	set_multiplayer_authority(Constants.SERVER_MULTIPLAYER_ID)
	player_input.set_multiplayer_authority(peer_id)
	rollback_synchronizer.process_authority()
```

`Player._rollback_tick()`
```
func _rollback_tick(delta: float, _tick: int, _is_fresh: bool) -> void:
	_force_update_is_on_floor()

	yaw.rotation_degrees.y = player_input.yaw_value
	pitch.rotation_degrees.x = player_input.pitch_value

	# Rotate character to yaw
	if player_input.is_rotating and player_input.input_dir != Vector2.ZERO:
		rotation.y = yaw.rotation.y

	var direction = (basis * Vector3(player_input.input_dir.x, 0, player_input.input_dir.y)).normalized()
	speed = RUNNING_SPEED if player_input.input_dir.y <= 0 else WALKING_SPEED
	if is_on_floor():
		set_grounded_velocity(delta, direction)
		if player_input.is_jumping:
			jump()
	else:
		set_airborne_velocity(delta, direction)

	# Apply movement
	velocity *= NetworkTime.physics_factor
	move_and_slide()
	velocity /= NetworkTime.physics_factor

	yaw.global_position = Vector3(
		global_position.x,
		global_position.y + 1.5,
		global_position.z
	)

func set_grounded_velocity(_delta: float, direction: Vector3) -> void:
	if direction:
		velocity.x = direction.x * speed
		velocity.z = direction.z * speed
	else:
		velocity.x = move_toward(velocity.x, 0, speed)
		velocity.z = move_toward(velocity.z, 0, speed)
```

`Player._force_update_is_on_floor()`
```
func _force_update_is_on_floor() -> void:
	var old_velocity = velocity
	velocity *= 0
	move_and_slide()
	velocity = old_velocity
```

I've confirmed that I have no `_physics_process` function defined and that my `_process` function simply deals with animation blending but here's what `_process` looks like for posterity
```
func _process(delta: float) -> void:
	var animation_tree = assassin_model.animation_tree
	var input_dir = player_input.input_dir
	var iwr_blend = animation_tree.get("parameters/iwr_blend/blend_amount")

	# Blend locomotion
	animation_vector = animation_vector.lerp(input_dir, 0.2)
	if input_dir == Vector2.ZERO:
		animation_tree.set("parameters/iwr_blend/blend_amount", lerpf(iwr_blend, -1.0, 1 - exp(-delta * 10.0)))
	elif input_dir.y <= 0:
		animation_tree.set("parameters/run/blend_position", animation_vector)
		animation_tree.set("parameters/iwr_blend/blend_amount", lerpf(iwr_blend, 1.0, 1 - exp(-delta * 10.0)))
	else:
		animation_tree.set("parameters/walk/blend_position", animation_vector)
		animation_tree.set("parameters/iwr_blend/blend_amount", lerpf(iwr_blend, 0.0, 1 - exp(-delta * 10.0)))

	# Handle blend for is_on_floor
	var on_floor_blend_target = 1.0 if is_on_floor() else 0.0
	var on_floor_blend = animation_tree.get("parameters/on_floor_blend/blend_amount")
	animation_tree.set(
		"parameters/on_floor_blend/blend_amount",
		lerpf(on_floor_blend, on_floor_blend_target, 10 * delta)
	)

	if player_input.is_stabbing:
		if player_input.input_dir == Vector2.ZERO:
			assassin_model.animation_player.play("stab")
		else:
			assassin_model.animation_tree.set(
				"parameters/stab/request",
				AnimationNodeOneShot.ONE_SHOT_REQUEST_FIRE
			)
```

Not sure exactly how else to look into this. Any guidance would be very appreciated. Thanks!

**Reply by elementbound** [reply]
Hi @jeremyhcho! I've been busier than usual recently, but I'll try to take a deeper look in the following days.

In the meanwhile, if it doesn't conflict with other systems in your game, you can reduce the physics tickrate, so sync to physics can be active without causing too much network traffic.

In the meanwhile, if you could attach a Godot project that replicates this issue, I'd be happy to look at that, it would probably simplify debugging the issue.

**Reply by jeremyhcho** [reply]
No worries! Thanks for the offer.

The interesting bit is that I can't seem to reproduce this issue on a new project but when I slim my current project down to what I think is pretty bare bones I still see this issue. I have a private repo that I don't mind inviting you to for debugging if you're open to helping me out

**Reply by elementbound** [reply]
For sure @jeremyhcho, I'd be happy to look at the repo!

**Reply by elementbound** [reply]
Hey @jeremyhcho! I'm still down to debug your repo if you'd like. Don't want to nag, just checking in :slightly_smiling_face:
